From 5e8dd3e1f10972aa4f4b714928a2e78038deff85 Mon Sep 17 00:00:00 2001
From: Jian-Ming Wang <jian.ming.wang@mitaccomputing.com>
Date: Mon, 2 Mar 2026 01:13:26 +0000
Subject: [PATCH 1/1] Use MiOBMC's naming convention to preset EntityID and
 EntityInstance when there is no definition for the sensor in EM
 configuration.

Upstream-Status: Inappropriate [No upstream maintenance]
---
 dbus-sdr/sdrutils.cpp | 70 +++++++++++++++++++++++++++++++++++++++++--
 1 file changed, 68 insertions(+), 2 deletions(-)

diff --git a/dbus-sdr/sdrutils.cpp b/dbus-sdr/sdrutils.cpp
index 0fd14c9..ea39606 100644
--- a/dbus-sdr/sdrutils.cpp
+++ b/dbus-sdr/sdrutils.cpp
@@ -502,6 +502,18 @@ const std::string* getSensorConfigurationInterface(
     return nullptr;
 }
 
+std::vector<std::string> splitString(const std::string& str, char delimiter) {
+    std::vector<std::string> tokens;
+    std::stringstream ss(str);
+    std::string token;
+
+    while (std::getline(ss, token, delimiter)) {
+        tokens.push_back(token);
+    }
+
+    return tokens;
+}
+
 // Follow Association properties for Sensor back to the Board dbus object to
 // check for an EntityId and EntityInstance property.
 void updateIpmiFromAssociation(
@@ -602,8 +614,62 @@ void updateIpmiFromAssociation(
 
         if (!sensorInterfacesResponseOpt.has_value())
         {
-            lg2::debug("Failed to GetObject, path: {PATH}", "PATH",
-                       sensorConfigPath);
+            // When single type of record in EM configuration generate multiple sensors,
+            // it will enter this case. Then we use MiOBMC's sensor naming convention to fill EntityID and Entity Instance.
+            std::vector<std::string> sensorNamingConv = splitString(sensorNameFromPath, '_');
+            if (sensorNamingConv.size() < 4)
+                continue;
+
+            if (sensorNamingConv[0] == "CPU")
+            {
+                entityId = 3;
+            }
+            else if (sensorNamingConv[0] == "GPU")
+            {
+                entityId = 43;
+            }
+            else if (sensorNamingConv[0] == "PSU")
+            {
+                entityId = 10;
+            }
+            else if (sensorNamingConv[0] == "DCSCM")
+            {
+                entityId = 6;
+            }
+            else if (sensorNamingConv[0] == "DIMM")
+            {
+                entityId = 32;
+            }
+            else if ((sensorNamingConv[0] == "Inlet") && sensorNameFromPath.contains("_TEMP_C"))
+            {
+                entityId = 55;
+            }
+            else
+            {
+                if (sensorNameFromPath.contains("Inlet") && sensorNameFromPath.contains("_TEMP_C"))
+                {
+                    entityId = 55;
+                    entityInstance = 0;
+                }
+                continue;
+            }
+
+            try
+            {
+                entityInstance = std::stoi(sensorNamingConv[1]);
+                lg2::debug("MiOBMC Naming Convention for {PATH}: EntityId = {ENTITYID}, EntityInstance = {ENTITYINSTANCE}",
+                                "PATH", sensorNameFromPath,
+                                "ENTITYID", entityId,
+                                "ENTITYINSTANCE", entityInstance);
+            }
+            catch (std::invalid_argument const& ex)
+            {
+                continue;
+            }
+            catch (std::out_of_range const& ex)
+            {
+                continue;
+            }
             continue;
         }
 
-- 
2.45.2

